src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs

```diff
@@ -121,11 +124,93 @@ protected override void RemoveItem(int index)
 
             base.RemoveItem(index);
 
+            if (!_skipRaisingEvents)
+            {
+                OnCountPropertyChanged();
+                OnIndexerPropertyChanged();
+                OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItem, index);
+            }
+        }
+
+        /// <summary>
+        /// Called by base class Collection&lt;T&gt; when a count of items is removed from the list;
+        /// raises a CollectionChanged event to any listeners.
+        /// </summary>
+        protected override void RemoveItemsRange(int index, int count)
+        {
+            CheckReentrancy();
+
+            int length = this.Count;
+            if (index < this.Count - count)
+            {
+                length = index + count;
+            }
+
+            List<T> removedItems = new List<T>();
```

Comment:
I agree with you, but the problem is the OnCollectionChanged requires an `IList` and that is all we use `removedItems` for. I believe it is better to create 1 List like we are doing rather than create an array and then copy that array to a `List<T>`

Reply:
> I believe it is better to create 1 List like we are doing rather than create an array and then copy that array to a `List<T>`

There's no need to copy the created array to a `List<T>`, arrays already implement `IList`.

Reply:
Something like:

```c#
T[] removedItems = new T[count];
for (int i = 0; i < count; i++)
{
    removedItems[i] = this[index + i];
}
```

We could also consider an optimization when `Items` is `List<T>` (which it likely will be the vast majority of the time, as it's the default):

```c#
T[] removedItems = new T[count];
if (Items is List<T> list)
{
    list.CopyTo(index, removedItems, 0, count);
}
else
{
    for (int i = 0; i < count; i++)
    {
        removedItems[i] = this[index + i];
    }
}
```

Reply:
This makes sense

Reply:
After running many tests, the proposed solution was not working on edge cases where we try and break the API where `length > int.MaxValue`. However, you did provide enough context so we can get it working.

Here is the entire method for context:
```c#
protected override void RemoveItemsRange(int index, int count)
{
    CheckReentrancy();

    int length = this.Count;
    if (index < this.Count - count)
    {
        length = index + count;
    }

    T[] oldItems = new T[Items.Count];
    if (Items is List<T> list)
    {
        list.CopyTo(oldItems);
    }
    else
    {
        for (int i = 0; i < oldItems.Length; i++)
        {
            oldItems[i] = this[i];
        }
    }

    bool ignore = _skipRaisingEvents;
    if (!ignore)
    {
        _skipRaisingEvents = true;
    }

    try
    {
        base.RemoveItemsRange(index, count);
    }
    finally
    {
        if (!ignore)
        {
            _skipRaisingEvents = false;
        }
    }
            
    if (count > 0 && !_skipRaisingEvents)
    {
        T[] removedItems = new T[length];
        for (int i = 0; i < count; i++)
        {
            removedItems[i] = oldItems[index + i];
        }

        OnCountPropertyChanged();
        OnIndexerPropertyChanged();
        OnCollectionChanged(NotifyCollectionChangedAction.Remove, removedItems, index);
    }
}
```

It looks like the changes you proposed in dotnet/coreclr#23166 would simplify this logic. I am going to update the logic with the proposed here and if we can get your changes merged and mirrored over quickly we can simplify

Reply:
This has been pushed, if we get dotnet/coreclr#23166 merged and mirrored in time, please open a new review so we can get this cleaned up